src/replay/consensus.zig

```diff
@@ -0,0 +1,1321 @@
+            replay_tower.markLastVoteTxBlockhashNonVoting();
+        },
+        else => {
+            // Do nothing
+        },
+    }
+}
+
+/// Processes a new root slot by updating various system components to reflect the new root.
+///
+/// - Validates the new root exists and has a hash.
+/// - Gets all slots rooted at the new root.
+/// - Updates ledger state with the new roots.
+/// - Cleans up progress map for non-existent slots.
+/// - Updates fork choice with the new root.
+///
+/// Analogous to [check_and_handle_new_root](https://github.com/anza-xyz/agave/blob/ccdcdbe9b6ff7dbd583d2101fe57b7cc41a6f863/core/src/replay_stage.rs#L4002)
+fn checkAndHandleNewRoot(
+    allocator: std.mem.Allocator,
+    blockstore_reader: *BlockstoreReader,
+    slot_tracker: *SlotTracker,
+    progress: *ProgressMap,
+    fork_choice: *ForkChoice,
+    new_root: Slot,
+) !void {
+    // get the root bank before squash.
+    if (slot_tracker.slots.count() == 0) return error.EmptySlotTracker;
+    var root_tracker = slot_tracker.slots.get(new_root) orelse return error.MissingSlot;
+    const maybe_root_hash, var hash_lg = root_tracker.state.hash.readWithLock();
+    defer hash_lg.unlock();
+    const root_hash = maybe_root_hash.* orelse return error.MissingHash;
+
+    const rooted_slots = try slot_tracker.parents(allocator, new_root);
+    defer allocator.free(rooted_slots);
+
+    // TODO implement leader_schedule_cache.set_root.
+    // TODO have this a seperate function?
+    {
+        var lowest_cleanup_slot = RwMux(Slot).init(0);
+        var max_root = std.atomic.Value(Slot).init(0);
+        var registry = sig.prometheus.Registry(.{}).init(allocator);
+        defer registry.deinit();
+
+        var writer = LedgerResultWriter{
+            .allocator = allocator,
+            .db = blockstore_reader.db,
+            .logger = .noop,
+            .lowest_cleanup_slot = &lowest_cleanup_slot,
+            .max_root = &max_root,
+            .scan_and_fix_roots_metrics = try registry.initStruct(
+                sig.ledger.result_writer.ScanAndFixRootsMetrics,
+            ),
+        };
+
+        try writer.setRoots(rooted_slots);
+    }
+
+    // Audit: The rest of the code maps to Self::handle_new_root in Agave.
+    // TODO
+    // - Prune program cache bank_forks.read().unwrap().prune_program_cache(new_root);
+    // - Set root on slot tracker: let removed_banks = bank_forks.write().unwrap().set_root(
+    //        new_root,
+    //        snapshot_controller,
+    //        highest_super_majority_root,
+    //    )?
+    //
+    // Update the progress map.
+    // TODO Move to its own function?
+    {
+        var to_remove = try std.ArrayListUnmanaged(Slot).initCapacity(
+            allocator,
+            progress.map.count(),
+        );
+        defer to_remove.deinit(allocator);
+
+        var it = progress.map.iterator();
+        while (it.next()) |entry| {
+            // TODO should frozen state be taking into consideration.
```

Comment:
If yes, then what? And do you know the answer? Maybe we should address this now.

Reply:
Cross checked and the Agave implementation, only keeps the slot in the progress map if it also found in the bank fork. Regardless of it is active or not.

[f3388ef](https://github.com/Syndica/sig/pull/762/commits/f3388ef590c6410198ea042433c6efe2b0f6b99d)